server/backends/chunkstore/chunkstore.go

```diff
@@ -462,65 +470,103 @@ func (l *writeLoop) write() error {
 	return err
 }
 
-func (l *writeLoop) readFromWriteChannel() {
+func (l *writeLoop) readFromWriteChannel() *WriteRequest {
 	l.timeout = false
-	var p []byte
 	select {
-	case p, l.open = <-l.writeChannel:
-		if l.open && p == nil {
-			l.flushTime = time.Unix(0, 0)
-		} else {
-			l.chunk = append(l.chunk, p...)
-			l.lastWriteSize = len(p)
+	case req, open := <-l.writeChannel:
+		l.open = open
+		if l.open {
+			if req.Chunk == nil {
+				l.flushTime = time.Unix(0, 0)
+			} else {
+				if req.VolatileTail != nil {
+					l.volatileTail = req.VolatileTail
+				}
+				l.chunk = append(l.chunk, req.Chunk...)
+				l.lastWriteSize = len(req.Chunk)
+			}
 		}
+		return &req
 	case <-time.After(time.Until(l.flushTime)):
 		l.timeout = true
+		return nil
 	case <-l.ctx.Done():
 		l.open = false
+		return nil
 	}
 }
 
 func (l *writeLoop) run() {
 	l.flushTime = time.Now().Add(year)
 	l.open = true
 	for l.open {
-		l.readFromWriteChannel()
+		req := l.readFromWriteChannel()
```

Comment:
nit: I think it would help readability & maintainability to have a few comments sprinkled in throughout this loop -- the logic here is pretty dense

Reply:
Hahaha, the readFromWriteChannel function used to just be in-line here, too 😶

I'll add more comments.